packages/nimble-components/src/select/index.ts

```diff
@@ -975,33 +990,96 @@ export class Select
      * @public
      */
     private filterOptions(): void {
-        const filter = this.filter.toLowerCase();
+        if (!this.$fastController.isConnected) {
```

Comment:
There is a lot going on in this function. I think it would be good to consider ways to make it more readable. Some ideas I had to consider are:

- Simplify the code path when `filter` is empty.
- Don't create a local function (particularly one this long/complex) within another function
- Split the logic that updates the state on an option/group from the logic that determines if it matches the filter.
- Create some helper functions that encapsulate more of this logic
- [Unsure] Put the logic that hides a group if it has no visible children in list-option-group itself. I'm not sure if that makes sense or not

I'd be happy to brainstorm more concrete ideas around this, but I thought I'd at least put a few possible directions here for you to consider.

Reply:
Refactored fairly heavily. Hopefully these concerns are addressed satisfactorily.